codec/address/bech32_codec.go

```diff
@@ -3,23 +3,92 @@ package address
 import (
 	"errors"
 	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/hashicorp/golang-lru/simplelru"
 
 	"cosmossdk.io/core/address"
 	errorsmod "cosmossdk.io/errors"
 
+	"github.com/cosmos/cosmos-sdk/internal/conv"
 	sdkAddress "github.com/cosmos/cosmos-sdk/types/address"
 	"github.com/cosmos/cosmos-sdk/types/bech32"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
+const (
+	// TODO: ideally sdk.GetBech32PrefixValAddr("") should be used but currently there's a cyclical import.
+	// 	Once globals are deleted the cyclical import won't happen.
+	prefixValAddr  = "valoper"
+	prefixConsAddr = "valcons"
+)
+
+// cache variables
+var (
+	accAddrMu     sync.Mutex
+	accAddrCache  *simplelru.LRU
+	consAddrMu    sync.Mutex
+	consAddrCache *simplelru.LRU
+	valAddrMu     sync.Mutex
+	valAddrCache  *simplelru.LRU
+
+	isCachingEnabled atomic.Bool
+)
+
+func init() {
+	var err error
+	isCachingEnabled.Store(true)
+
+	// in total the cache size is 61k entries. Key is 32 bytes and value is around 50-70 bytes.
+	// That will make around 92 * 61k * 2 (LRU) bytes ~ 11 MB
+	if accAddrCache, err = simplelru.NewLRU(60000, nil); err != nil {
```

Comment:
personal preference: it would be great to have this configurable so people can trade CPU for memory or not.

Reply:
Agreed. I'm not sure what the best approach is to solve this, as LRUs are currently shared between codecs and defined as globals.